File: platform/macos/Sources/TailScoutCore/TailscaleStatus.swift
import Foundation

public struct TailscaleStatus: Decodable, Equatable, Sendable {
    public let version: String
    public let clientVersion: String
    public let backendState: BackendState
    public let tun: Bool
    public let magicDNSSuffix: String
    public let currentTailnet: Tailnet?
    public let health: [String]
    public let thisNode: TailscaleNode?
    public let peers: [TailscaleNode]
    public let users: [UInt64: UserProfile]

    enum CodingKeys: String, CodingKey {
        case version = "Version"
        case clientVersion = "ClientVersion"
        case backendState = "BackendState"
        case tun = "TUN"
        case magicDNSSuffix = "MagicDNSSuffix"
        case currentTailnet = "CurrentTailnet"
        case health = "Health"
        case thisNode = "Self"
        case peer = "Peer"
        case user = "User"
    }

    public init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        version = try values.decodeString(forKey: .version)
        clientVersion = try values.decodeString(forKey: .clientVersion)
        backendState = BackendState(rawValue: try values.decodeString(forKey: .backendState))
        tun = try values.decodeBool(forKey: .tun)
        magicDNSSuffix = try values.decodeString(forKey: .magicDNSSuffix)
        currentTailnet = try values.decodeIfPresent(Tailnet.self, forKey: .currentTailnet)
        health = try values.decodeStringArray(forKey: .health)
        thisNode = try values.decodeIfPresent(TailscaleNode.self, forKey: .thisNode)
        let peerMap = try values.decodeIfPresent([String: TailscaleNode].self, forKey: .peer) ?? [:]
        peers = Array(peerMap.values)

        let rawUsers = try values.decodeIfPresent([String: UserProfile].self, forKey: .user) ?? [:]
        users = rawUsers.reduce(into: [:]) { result, entry in
            guard let id = UInt64(entry.key) else { return }
            var profile = entry.value
            profile.id = id
            result[id] = profile
        }
    }

    public static func parse(_ input: String) throws -> Self {
        try parse(Data(input.utf8))
    }

    public static func parse(_ data: Data) throws -> Self {
        try JSONDecoder().decode(Self.self, from: data)
    }

    public var displayVersion: String {
        version.isEmpty ? clientVersion : version
    }

    public var sortedPeers: [TailscaleNode] {
        peers.sorted {
            $0.online == $1.online
                ? $0.displayName.localizedCaseInsensitiveCompare($1.displayName) == .orderedAscending
                : $0.online
        }
    }

    public var exitNodeOptions: [TailscaleNode] {
        sortedPeers.filter(\.exitNodeOption)
    }

    public func ownerLabel(for node: TailscaleNode) -> String? {
        users[node.userID]?.displayLabel
    }

    public func hasSameOwner(as node: TailscaleNode) -> Bool {
        guard let thisNode else { return true }
        return thisNode.userID == 0 || node.userID == 0 || thisNode.userID == node.userID
    }

    public func canSendTaildrop(to node: TailscaleNode) -> Bool {
        node.canReceiveTaildrop && hasSameOwner(as: node)
    }
}

File: platform/windows/TailScout.Windows.Core/Models/TailscaleStatus.cs
namespace TailScout.Windows.Core.Models;

public sealed record TailscaleStatus(
    string Version,
    string ClientVersion,
    string BackendState,
    bool Tun,
    string MagicDnsSuffix,
    TailnetInfo? CurrentTailnet,
    IReadOnlyList<string> Health,
    TailNode? ThisNode,
    IReadOnlyList<TailNode> Peers,
    IReadOnlyDictionary<ulong, UserProfile> Users)
{
    public string StatusLabel => BackendState switch
    {
        "NeedsLogin" => "Logged out",
        "Running" => "Connected",
        "Stopped" => "Disconnected",
        "Starting" => "Starting…",
        "" => "Unknown",
        _ => BackendState
    };

    public string DisplayVersion => Text.First(Version, ClientVersion);
    public IReadOnlyList<TailNode> SortedPeers => Peers
        .OrderByDescending(peer => peer.Online)
        .ThenBy(peer => peer.DisplayName, StringComparer.OrdinalIgnoreCase)
        .ToArray();
    public string? OwnerLabel(TailNode node) =>
        Users.TryGetValue(node.UserId, out var user) ? user.DisplayLabel : null;

    public bool HasSameOwner(TailNode node) =>
        ThisNode is null || ThisNode.UserId == 0 || node.UserId == 0 || ThisNode.UserId == node.UserId;

    public bool CanSendTaildropTo(TailNode node) => node.CanReceiveTaildrop && HasSameOwner(node);

    public static TailscaleStatus Parse(string json)
    {
        var raw = JsonModel.Parse<RawStatus>(json);
        var users = (raw.User ?? []).Select(entry =>
        {
            var id = ulong.TryParse(entry.Key, out var keyId) ? keyId : entry.Value.ID ?? 0;
            var value = entry.Value;
            return new UserProfile(id, Text.Value(value.LoginName), Text.Value(value.DisplayName));
        }).Where(user => user.Id > 0).ToDictionary(user => user.Id);

        return new(
            Text.Value(raw.Version), Text.Value(raw.ClientVersion), Text.Value(raw.BackendState),
            raw.TUN ?? false, Text.Value(raw.MagicDNSSuffix), raw.CurrentTailnet is null ? null : new(
                Text.Value(raw.CurrentTailnet.Name), Text.Value(raw.CurrentTailnet.MagicDNSSuffix),
                raw.CurrentTailnet.MagicDNSEnabled ?? false), raw.Health ?? [],
            raw.Self is null ? null : TailNode.From(raw.Self),
            (raw.Peer ?? []).Values.Select(TailNode.From).ToArray(), users);
    }
}

internal sealed class RawStatus
{
    public string? Version { get; init; }
    public string? ClientVersion { get; init; }
    public string? BackendState { get; init; }
    public bool? TUN { get; init; }
    public string? MagicDNSSuffix { get; init; }
    public RawTailnet? CurrentTailnet { get; init; }
    public string[]? Health { get; init; }
    public RawNode? Self { get; init; }
    public Dictionary<string, RawNode>? Peer { get; init; }
    public Dictionary<string, RawUser>? User { get; init; }
}

internal sealed class RawTailnet
{
    public string? Name { get; init; }
    public string? MagicDNSSuffix { get; init; }
    public bool? MagicDNSEnabled { get; init; }
}

internal sealed class RawUser
{
    public ulong? ID { get; init; }
    public string? LoginName { get; init; }
    public string? DisplayName { get; init; }
}
